setup.py

```diff
@@ -10,24 +10,99 @@
 # SPDX-License-Identifier: ISC
 
 import setuptools
+import os
+import re
+import sys
+import platform
+import subprocess
+
+from distutils.version import LooseVersion
+from setuptools import Extension
+from setuptools.command.build_ext import build_ext
 
 with open("README.md", "r") as fh:
     long_description = fh.read()
 
+
+# https://www.benjack.io/2017/06/12/python-cpp-tests.html
+class CMakeExtension(Extension):
+    def __init__(self, name, sourcedir='', prefix=''):
+        Extension.__init__(self, name, sources=[])
+        self.sourcedir = os.path.abspath(sourcedir)
+        self.prefix = prefix
+
+
+class CMakeBuild(build_ext):
+    def run(self):
+        try:
+            out = subprocess.check_output(['cmake', '--version'])
+        except OSError:
+            raise RuntimeError(
+                "CMake must be installed to build the following extensions: " +
+                ", ".join(e.name for e in self.extensions))
+
+        if platform.system() == "Windows":
+            cmake_version = LooseVersion(
+                re.search(r'version\s*([\d.]+)', out.decode()).group(1))
+            if cmake_version < '3.1.0':
+                raise RuntimeError("CMake >= 3.1.0 is required on Windows")
+
+        for ext in self.extensions:
+            self.build_extension(ext)
+
+    def build_extension(self, ext):
+        extdir = os.path.join(
+            os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name))),
+            ext.prefix)
+        cmake_args = [
+            '-DCMAKE_INSTALL_PREFIX=' + extdir,
+            '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
+            '-DPYTHON_EXECUTABLE=' + sys.executable
+        ]
+
+        cfg = 'Debug' if self.debug else 'Release'
+        build_args = ['--config', cfg]
+
+        if platform.system() == "Windows":
+            cmake_args += [
+                '-DCMAKE_INSTALL_PREFIX=' + extdir,
+                '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}'.format(
+                    cfg.upper(), extdir)
+            ]
+            if sys.maxsize > 2**32:
```

Comment:
Can you please add a comment on what is going on here?  Should this logic be located within the `CMakeLists.txt` rather than here?

Reply:
This was based on [an article](https://www.benjack.io/2018/02/02/python-cpp-revisited.html) I found. [(code)](https://github.com/benjaminjack/python_cpp_example/blob/master/setup.py)

I don't have a Windows machine to test this on, but this is specific to that platform, so I thought I'd just leave it in.

Reply:
If the Windows stuff is untested, I'd lean to just leaving it out?  We can always re-add this stuff once we work on packaging for Windows PyPi?

Reply:
Done.